skglm/solvers/prox_newton.py

```diff
@@ -0,0 +1,351 @@
+import numpy as np
+from numba import njit
+from scipy.sparse import issparse
+
+
+EPS_TOL = 0.3
+MAX_CD_ITER = 20
+MAX_BACKTRACK_ITER = 20
+
+
+def prox_newton(X, y, datafit, penalty, w_init=None, p0=10,
+                max_iter=20, max_epochs=1000, tol=1e-4, verbose=0):
+    """Run a Prox Newton solver combined with working sets.
+
+    Parameters
+    ----------
+    X : array or sparse CSC matrix, shape (n_samples, n_features)
+        Design matrix.
+
+    y : array, shape (n_samples,)
+        Target vector.
+
+    datafit : instance of BaseDatafit
+        Datafit object.
+
+    penalty : instance of BasePenalty
+        Penalty object.
+
+    w_init : array, shape (n_features,), default None
+        Initial value of coefficients.
+        If set to None, a zero vector is used instead.
+
+    p0 : int, default 10
+        Minimum number of features to be included in the working set.
+
+    max_iter : int, default 20
+        Maximum number of outer iterations.
+
+    max_epochs : int, default 1000
+        Maximum number of CD epochs to find the descent direction.
+
+    tol : float, default 1e-4
+        Tolerance for convergence.
+
+    verbose : bool, default False
+        Amount of verbosity. 0/False is silent.
+
+    Returns
+    -------
+    w : array, shape (n_features,)
+        Solution that minimizes the problem defined by datafit and penalty.
+
+    objs_out: array (max_iter,)
+        The objective values at every outer iteration.
+
+    stop_crit: float
+        The value of the stopping criterion when the solver stops.
+
+    References
+    ----------
+    .. [1] M. Massias, A. Gramfort, J. Salmon
```

Comment:
Celer-PN is detailed in the JMLR paper : https://arxiv.org/abs/1907.05830 (put this link as there is an error in the JMLR website version)